modyn/trainer_server/internal/trainer/remote_downsamplers/deepcore_utils/orthogonal_matching_pursuit.py

```diff
@@ -0,0 +1,105 @@
+# flake8: noqa
+# mypy: ignore-errors
+
+import numpy as np
+import torch
+from scipy.linalg import lstsq
+from scipy.optimize import nnls
+
+
+def orthogonal_matching_pursuit(A, b, budget: int, lam: float = 1.0):
+    """approximately solves min_x |x|_0 s.t. Ax=b using Orthogonal Matching Pursuit
+    Acknowlegement to:
+    https://github.com/krishnatejakk/GradMatch/blob/main/GradMatch/selectionstrategies/helpers/omp_solvers.py
+    Args:
+      A: design matrix of size (d, n)
+      b: measurement vector of length d
+      budget: selection budget
+      lam: regularization coef. for the final output vector
+    Returns:
+       vector of length n
+    """
+    with torch.no_grad():
+        d, n = A.shape
+        if budget <= 0:
+            budget = 0
+        elif budget > n:
+            budget = n
+
+        x = np.zeros(n, dtype=np.float32)
+        resid = b.clone()
+        indices = []
+        boolean_mask = torch.ones(n, dtype=bool, device="cuda")
+        all_idx = torch.arange(n, device="cuda")
+
+        for i in range(budget):
+            projections = torch.matmul(A.T, resid)
+            index = torch.argmax(projections[boolean_mask])
+            index = all_idx[boolean_mask][index]
+
+            indices.append(index.item())
+            boolean_mask[index] = False
+
+            if indices.__len__() == 1:
+                A_i = A[:, index]
+                x_i = projections[index] / torch.dot(A_i, A_i).view(-1)
+                A_i = A[:, index].view(1, -1)
+            else:
+                A_i = torch.cat((A_i, A[:, index].view(1, -1)), dim=0)
+                temp = torch.matmul(A_i, torch.transpose(A_i, 0, 1)) + lam * torch.eye(A_i.shape[0], device="cuda:0")
```

Comment:
Adapted DeepCore's version since lstsq has been updated in Pytorch. See [here](https://github.com/PatrickZH/DeepCore/issues/10)